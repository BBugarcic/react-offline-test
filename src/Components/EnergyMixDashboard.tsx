import { Result, Spin } from 'antd'
import 'antd/dist/antd.css'
import fetch from 'node-fetch'
import React, { Component } from 'react'
import Moment from 'react-moment'
import '../Styles/styles.css'
import FuelPerc from './FuelPerc'

interface EnergyMixProps {}

export interface EnergyMixState {
  error?: any
  isLoaded: boolean
  data?: Data
}

export interface Data {
  from: string
  to: string
  generationmix: FuelPercTotal[]
}

interface FuelPercTotal {
  fuel: string
  perc: number
}

class EnergyMixDashboard extends Component<EnergyMixProps, EnergyMixState> {
  constructor(props) {
    super(props)
    this.state = {
      error: null,
      isLoaded: true,
      data: undefined,
    }
  }

  componentDidMount() {
    fetch('https://api.carbonintensity.org.uk/generation')
      .then(res => res.json())
      .then(
        result => {
          this.setState({
            isLoaded: true,
            data: result.data,
          })
        },
        error => {
          this.setState({
            isLoaded: true,
            error,
          })
        }
      )
  }

  render() {
    const { isLoaded, error, data } = this.state
    const generationMix: FuelPercTotal[] =
      data !== undefined ? data.generationmix : null

    if (error) {
      return <Result status='error' title='Error' subTitle={error.message} />
    } else if (!isLoaded) {
      return <Spin size='large' />
    } else {
      return (
        <>
          <Period
            startTime={data ? data.from : null}
            endTime={data ? data.to : null}
          />
          <div className='dashboard'>
            {generationMix !== null &&
              generationMix.map((item: FuelPercTotal, index: number) => {
                return (
                  <FuelPerc
                    key={`${index}-${name}`}
                    name={item.fuel}
                    perc={item.perc}
                  />
                )
              })}
          </div>
        </>
      )
    }
  }
}

export default EnergyMixDashboard

interface PeriodProps {
  startTime: string
  endTime: string
}

export const Period: React.FC<PeriodProps> = ({ startTime, endTime }) => {
  return (
    <h3 className='left-align-title no-margin'>
      Setelment period: &nbsp;
      <Moment format='DD-MM-YYYY LT'>{startTime}</Moment>&nbsp;-&nbsp;
      <Moment format='LT'>{endTime}</Moment>
    </h3>
  )
}
