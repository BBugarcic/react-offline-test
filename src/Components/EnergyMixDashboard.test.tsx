import { mount } from 'enzyme'
import React from 'react'
import EnergyMixDashboard, { Data } from './EnergyMixDashboard'

const data: Data = {
  from: '2019-08-12T12:30Z',
  to: '2019-08-12T13:00Z',
  generationmix: [
    {
      fuel: 'biomass',
      perc: 4.8,
    },
    {
      fuel: 'coal',
      perc: 2.5,
    },
    {
      fuel: 'imports',
      perc: 8.7,
    },
    {
      fuel: 'gas',
      perc: 46.5,
    },
    {
      fuel: 'nuclear',
      perc: 16.1,
    },
    {
      fuel: 'other',
      perc: 0.3,
    },
    {
      fuel: 'hydro',
      perc: 0.9,
    },
    {
      fuel: 'solar',
      perc: 14.6,
    },
    {
      fuel: 'wind',
      perc: 5.6,
    },
  ],
}

describe('Dashboard', () => {
  it('should render error message when there is no data provided', () => {
    const component = mount(<EnergyMixDashboard />)
    component.setState({
      isLoaded: true,
      data: undefined,
      error: { message: 'Failed to fetch' },
    })
    expect(component).toMatchSnapshot()
  })
})

describe('Dashboard', () => {
  it('should render EnergyMixDashboard when there is data provided', () => {
    const component = mount(<EnergyMixDashboard />)
    component.setState({
      isLoaded: true,
      data: data,
      error: null,
    })
    expect(component).toMatchSnapshot()
  })
})

describe('Dashboard', () => {
  it('should render spinner if data is not loaded', () => {
    const component = mount(<EnergyMixDashboard />)
    component.setState({
      isLoaded: false,
      data: data,
      error: null,
    })
    expect(component).toMatchSnapshot()
  })
})
