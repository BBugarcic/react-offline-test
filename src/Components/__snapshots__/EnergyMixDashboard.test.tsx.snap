// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should render spinner if data is not loaded 1`] = `
<EnergyMixDashboard>
  <Spin
    size="large"
    spinning={true}
    wrapperClassName=""
  >
    <div
      className="ant-spin ant-spin-lg ant-spin-spinning"
    >
      <span
        className="ant-spin-dot ant-spin-dot-spin"
      >
        <i
          className="ant-spin-dot-item"
        />
        <i
          className="ant-spin-dot-item"
        />
        <i
          className="ant-spin-dot-item"
        />
        <i
          className="ant-spin-dot-item"
        />
      </span>
    </div>
  </Spin>
</EnergyMixDashboard>
`;
